### PR TITLE
[proto] Speed improvements for adjust hue op

### PR DESCRIPTION
**Pull request by vfdev-5**
Results:
```
[--------------- Adjust_hue cpu torch.uint8 --------------]
                     |  adjust_hue stable  |  adjust_hue v2
1 threads: ------------------------------------------------
      (3, 400, 400)  |         18.9        |       13.9    
6 threads: ------------------------------------------------
      (3, 400, 400)  |          5.4        |        4.5    

Times are in milliseconds (ms).

[-------------- Adjust_hue cuda torch.uint8 --------------]
                     |  adjust_hue stable  |  adjust_hue v2
1 threads: ------------------------------------------------
      (3, 400, 400)  |         1829        |       671     
6 threads: ------------------------------------------------
      (3, 400, 400)  |         1477        |       673     

Times are in microseconds (us).

[-------------- Adjust_hue cpu torch.float32 -------------]
                     |  adjust_hue stable  |  adjust_hue v2
1 threads: ------------------------------------------------
      (3, 400, 400)  |         18.3        |       13.4    
6 threads: ------------------------------------------------
      (3, 400, 400)  |          5.4        |        4.4    

Times are in milliseconds (ms).

[------------- Adjust_hue cuda torch.float32 -------------]
                     |  adjust_hue stable  |  adjust_hue v2
1 threads: ------------------------------------------------
      (3, 400, 400)  |         1441        |       630     
6 threads: ------------------------------------------------
      (3, 400, 400)  |         1447        |       636     

Times are in microseconds (us).

```

[Source](https://github.com/vfdev-5/tvapiv2_benchmarks/blob/145e951f9f0b00dccd6077e7925ea1adeadc027b/output/20221021-141927-output-adjust-hue_6bde2919b.log), [code](https://github.com/vfdev-5/tvapiv2_benchmarks/blob/0d081204477a88dd3ee4d887437bf59faf4f0f3c/check_adjust_hue.py)

### CODE REVIEW

File: torchvision/prototype/transforms/functional/_color.py [hunk 1]
```diff
@@ -98,7 +98,98 @@ def adjust_sharpness(inpt: features.InputTypeJIT, sharpness_factor: float) -> fe
...
+
+    if c not in [1, 3]:
+        raise TypeError(f"Input image tensor permitted channel values are {[1, 3]}, but found {c}")
+
+    if c == 1:  # Match PIL behaviour
+        return image
+
+    orig_dtype = image.dtype
+    if image.dtype == torch.uint8:
+        image = image.to(dtype=torch.float32) / 255.0
```
Comment (datumbox) [inline torchvision/prototype/transforms/functional/_color.py:179]: Wouldn't be faster to omit the `to()` here and only rely on the devision to cast to floats?

**Review by datumbox — APPROVED** [review]
LGTM, just 2 minor comments.

File: torchvision/prototype/transforms/functional/_color.py [hunk 2]
```diff
@@ -98,7 +98,98 @@ def adjust_sharpness(inpt: features.InputTypeJIT, sharpness_factor: float) -> fe
         return adjust_sharpness_image_pil(inpt, sharpness_factor=sharpness_factor)
 
 
-adjust_hue_image_tensor = _FT.adjust_hue
+def _rgb_to_hsv(image: torch.Tensor) -> torch.Tensor:
+    r, g, _ = image.unbind(dim=-3)
+
+    # Implementation is based on https://github.com/python-pillow/Pillow/blob/4174d4267616897df3746d315d5a2d0f82c656ee/
+    # src/libImaging/Convert.c#L330
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_color.py:150]: Please put this on one line to be able to copy the link

```suggestion
    # Implementation is based on 
    # https://github.com/python-pillow/Pillo/blob/4174d4267616897df3746d315d5a2d0f82c656ee/src/libImaging/Convert.c#L330
```

File: torchvision/prototype/transforms/functional/_color.py [hunk 3]
```diff
@@ -98,7 +98,98 @@ def adjust_sharpness(inpt: features.InputTypeJIT, sharpness_factor: float) -> fe
...
+    # from happening in the results, because
+    #   + S channel has division by `maxc`, which is zero only if `maxc = minc`
+    #   + H channel has division by `(maxc - minc)`.
+    #
+    # Instead of overwriting NaN afterwards, we just prevent it from occuring so
+    # we don't need to deal with it in case we save the NaN in a buffer in
+    # backprop, if it is ever supported, but it doesn't hurt to do so.
+    eqc = maxc == minc
+
+    cr = maxc - minc
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_color.py:163]: Could you make these variable names more expressive? It is hard to understand what is actually happening if the values are hidden behind acronyms. I'm totally ok with `r`, `g`, and `b` and in turn with `h`, `s`, and `v`. But `eqc`, `cr`, `rc`, ... are opaque. I'm aware that PIL uses the same names, but I would prefer to have expressive names in our implementation and maybe a comment stating the corresponding PIL variable.

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_color.py:163]: The point of this PR is to optimize the runtime reusing existing code. Some vars are just temporary vars. Others have a certain meaning (to me): cr = channel_range. I'll replace cr with `channel_range`, for others I do not have a good name.

Comment (pmeier) [reply torchvision/prototype/transforms/functional/_color.py:163]: `cr` to `channel_range` is definitely an improvement!

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

**Pull request merged**